pkg/sloop/queries/timerange.go

```diff
@@ -8,60 +8,152 @@
 package queries
 
 import (
+	"fmt"
 	"github.com/golang/glog"
 	"github.com/golang/protobuf/ptypes"
 	"github.com/salesforce/sloop/pkg/sloop/store/typed"
 	"github.com/salesforce/sloop/pkg/sloop/store/untyped"
 	"net/url"
+	"strconv"
 	"time"
 )
 
-// This computes a start and end time for a given query.  If user specifies a time duration we use it, otherwise
-// we use maxLookBack from config.  For the end time if not specified we want to use newest data in the store.
-// That way we can look at old data sets without clipping.  We know the min and max time of the newest partition
-// but we dont really know the newest record within that range.  For now just use end of newest partiton, but
-// we will improve that later.
-// TODO: Add unit tests
+const minLookback = 1 * time.Minute
+
+// This computes a start and end time for a given query.  There are 2 ways this can be specified:
+//
+// Using "lookback":
+//
+//   We first find the endTime.  If we are looking at historic data, we use the end of the last partitions.  If
+//   that is in the future, we use now().  We don't want to always use now() as that would prevent users from looking
+//   at old data as it would get clipped by maxLookBack
+//   StartTime is just endTime - lookback
+//
+// Using "start_time" and "end_time"
+//
+//   This is straight forward.  These are UTC Unix times
+//
 // TODO: If wall clock is in the middle of the newest partition min-max time we can use it
-func computeTimeRange(params url.Values, tables typed.Tables, maxLookBack time.Duration) (time.Time, time.Time) {
-	now := time.Now()
+func computeTimeRange(params url.Values, tables typed.Tables, maxLookBack time.Duration) (time.Time, time.Time, error) {
+	endOfTime := getEndOfTime(tables)
+	return computeTimeRangeInternal(params, endOfTime, maxLookBack)
+}
+
+func computeTimeRangeInternal(params url.Values, endOfTime time.Time, maxLookBack time.Duration) (time.Time, time.Time, error) {
+	lookBackVal := params.Get(LookbackParam)
+	startTimeVal := params.Get(StartTimeParam)
+	endTimeVal := params.Get(EndTimeParam)
+
+	var computedStart time.Time
+	var computedEnd time.Time
+	var err error
+
+	// Input validations
+	if startTimeVal == "" && endTimeVal == "" && lookBackVal == "" {
+		return time.Time{}, time.Time{}, fmt.Errorf("Time range must be set with either [%v] or both of [%v,%v] but all 3 were empty", LookbackParam, StartTimeParam, EndTimeParam)
+	}
+	if lookBackVal != "" {
+		if startTimeVal != "" || endTimeVal != "" {
+			return time.Time{}, time.Time{}, fmt.Errorf("When [%v] is set, you can not set either of [%v,%v].  Got (%v,%v,%v) respectively", LookbackParam, StartTimeParam, EndTimeParam, lookBackVal, startTimeVal, endTimeVal)
+		}
+	} else {
+		if (startTimeVal == "") || (endTimeVal == "") {
+			return time.Time{}, time.Time{}, fmt.Errorf("Either %v and %v both need to be set or neither set.  Got (%v,%v) respectively", StartTimeParam, EndTimeParam, startTimeVal, endTimeVal)
```

Comment:
nit: error message needs to be updated (`Both params [%v,%v] must be set`)

Reply:
meh, I don't think anyone would be confused by this error